src/transport.rs

```diff
@@ -0,0 +1,473 @@
+use std::{
+    collections::HashMap,
+    pin::Pin,
+    str::FromStr,
+    task::{Context, Poll, Waker},
+};
+use tracing::debug;
+
+use crate::connection::{Connection, InnerConnection, PendingConnection};
+use crate::error::NymTransportError;
+use crate::message::{ConnectionId, ConnectionMessage, InboundMessage, Message, OutboundMessage};
+use crate::mixnet::Mixnet;
+
+/// NymTransport implements the Transport trait using the Nym mixnet.
+pub struct NymTransport {
+    // our Nym address
+    self_address: Recipient,
+    pub(crate) listen_addr: Multiaddr,
+    pub(crate) listener_id: ListenerId,
+
+    // established connections
+    connections: HashMap<ConnectionId, InnerConnection>,
+
+    // outbound pending dials
+    pending_dials: HashMap<ConnectionId, PendingConnection>,
+
+    // inbound mixnet messages
+    inbound_rx: Receiver<InboundMessage>,
+
+    // outbound mixnet messages
+    outbound_tx: Sender<OutboundMessage>,
+
+    // inbound messages for Transport.poll()
+    poll_rx: Receiver<TransportEvent<Upgrade, NymTransportError>>,
+
+    // outbound messages to Transport.poll()
+    poll_tx: Sender<TransportEvent<Upgrade, NymTransportError>>,
+
+    mixnet: Mixnet,
+    waker: Option<Waker>,
+}
+
+impl NymTransport {
+    pub async fn new(uri: &String) -> Result<Self, Error> {
+        // accept websocket uri and call Mixnet::new()
+        // then, cache our Nym address and create the listener for it
+        let (mut mixnet, inbound_rx, outbound_tx) = Mixnet::new(uri).await?;
+        let self_address = mixnet.get_self_address().await?;
+        let listen_addr = nym_address_to_multiaddress(self_address)?;
+        let listener_id = ListenerId::new();
+
+        #[allow(clippy::type_complexity)]
+        let (poll_tx, poll_rx): (
+            Sender<TransportEvent<Upgrade, NymTransportError>>,
+            Receiver<TransportEvent<Upgrade, NymTransportError>>,
+        ) = async_channel::unbounded();
+
+        poll_tx
+            .send(TransportEvent::NewAddress {
+                listener_id,
+                listen_addr: listen_addr.clone(),
+            })
+            .await?;
+
+        Ok(Self {
+            self_address,
+            listen_addr,
+            listener_id,
+            connections: HashMap::<ConnectionId, InnerConnection>::new(),
+            pending_dials: HashMap::<ConnectionId, PendingConnection>::new(),
+            inbound_rx,
+            outbound_tx,
+            poll_rx,
+            poll_tx,
+            mixnet,
+            waker: None,
+        })
+    }
+
+    // handle_connection_response resolves the pending connection corresponding to the response
+    // (if there is one) into a Connection.
+    fn handle_connection_response(&mut self, msg: ConnectionMessage) -> Result<(), Error> {
+        let pending_conn = self.pending_dials.get(&msg.id);
+        if pending_conn.is_none() {
+            return Err(anyhow!("no connection found for ConnectionRespone"));
```

Comment:
This should be a variant of NymTransportError, ie: `NymTransportError::NoConnectionFound`